routers/api/v1/repo/branch.go

```diff
@@ -88,19 +101,204 @@ func ListBranches(ctx *context.APIContext) {
 	//     "$ref": "#/responses/BranchList"
 	branches, err := ctx.Repo.Repository.GetBranches()
 	if err != nil {
-		ctx.Error(500, "GetBranches", err)
+		ctx.Error(http.StatusInternalServerError, "GetBranches", err)
 		return
 	}
 
 	apiBranches := make([]*api.Branch, len(branches))
 	for i := range branches {
 		c, err := branches[i].GetCommit()
 		if err != nil {
-			ctx.Error(500, "GetCommit", err)
+			ctx.Error(http.StatusInternalServerError, "GetCommit", err)
+			return
+		}
+
+		protected, err := ctx.Repo.Repository.IsProtectedBranch(branches[i].Name, ctx.Repo.Owner)
+		if err != nil {
+			ctx.Error(http.StatusInternalServerError, "IsProtectedBranch", err)
+			return
+		}
+
+		apiBranches[i] = convert.ToBranch(ctx.Repo.Repository, branches[i], c, protected)
+	}
+
+	ctx.JSON(http.StatusOK, &apiBranches)
+}
+
+// GetProtectedBranchBy getting protected branch by ID/Name
+func GetProtectedBranchBy(ctx *context.APIContext) {
+	// swagger:operation GET /repos/{owner}/{repo}/branches/{branch}/protection repository repoGetBranchProtection
+	// ---
+	// summary: Retrieve a specific branch protection from a repository
+	// produces:
+	// - application/json
+	// parameters:
+	// - name: owner
+	//   in: path
+	//   description: owner of the repo
+	//   type: string
+	//   required: true
+	// - name: repo
+	//   in: path
+	//   description: name of the repo
+	//   type: string
+	//   required: true
+	// - name: branch
+	//   in: path
+	//   description: branch to get
+	//   type: string
+	//   required: true
+	// responses:
+	//   "200":
+	//     "$ref": "#/responses/Branch"
+	protectBranch, err := models.GetProtectedBranchBy(ctx.Repo.Repository.ID, ctx.Params(":branch"))
+	if err != nil {
+		if !git.IsErrBranchNotExist(err) {
+			ctx.Error(http.StatusInternalServerError, "GetProtectedBranchBy", err)
 			return
 		}
```

Comment:
else (other errors?)

Reply:
Also, xorm will not give an error if there is no row in the database, but an empty object. So you have to check protectBranch against nil.